File: minWindowSubstr.js
function minWindow(s, t) {
    const charCount = new Map();
    for (const char of t) {
        charCount.set(char, (charCount.get(char) || 0) + 1);
    }

    let left = 0;
    let minLen = Infinity;
    let minLenStart = 0;
    let count = t.length;

    for (let right = 0; right < s.length; right++) {
        const rightChar = s[right];
        if (charCount.has(rightChar)) {
            charCount.set(rightChar, charCount.get(rightChar) - 1);
            if (charCount.get(rightChar) >= 0) {
                count--;
            }
        }

        while (count === 0) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                minLenStart = left;
            }
            const leftChar = s[left];
            if (charCount.has(leftChar)) {
                charCount.set(leftChar, charCount.get(leftChar) + 1);
                if (charCount.get(leftChar) > 0) {
                    count++;
                }
            }
            left++;
        }
    }

    return minLen === Infinity ? "" : s.substring(minLenStart, minLenStart + minLen);
}

const s = "ADOBECODEBANC";
const t = "ABC";
console.log(minWindow(s, t)); // Output: "BANC"

File: minWindowSubstr.go
package main

import "math"

func minWindow(s string, t string) string {
	charCount := make(map[byte]int)
	for _, char := range t {
		charCount[byte(char)]++
	}

	left, minLen, minLenStart, count := 0, math.MaxInt32, 0, len(t)

	for right := 0; right < len(s); right++ {
		rightChar := s[right]
		if charCount[rightChar] > 0 {
			count--
		}
		charCount[rightChar]--

		for count == 0 {
			if right-left+1 < minLen {
				minLen = right - left + 1
				minLenStart = left
			}
			leftChar := s[left]
			charCount[leftChar]++
			if charCount[leftChar] > 0 {
				count++
			}
			left++
		}
	}

	if minLen == math.MaxInt32 {
		return ""
	}
	return s[minLenStart : minLenStart+minLen]
}
